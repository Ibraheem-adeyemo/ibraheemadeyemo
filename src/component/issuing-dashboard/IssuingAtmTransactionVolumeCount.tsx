import React, { useContext, useEffect, useState } from "react";
import { Box, Text, Flex, HStack, Tag } from "@chakra-ui/react";
import {
  IssuingLineChart,
  IssuingBarChart,
  IssuingBarChartHorizontal,
  IssuingBarLineChart,
  IssuingLineChartSingle,
} from "../app-charts";
import { chartContainerSx } from "../../sx";
import { AuthContext, StatsContext } from "../../providers";
import { getUrlForSuperadminORBankAdmin } from "../../lib";
import { apiUrlsv1, appRoles } from "../../constants";
import useSWR from "swr";
import { IssuingFaileSuccessProps } from "../../models/issuing-dashboard";
import { data2 } from ".";
import { sumBy } from "lodash";

const containerType = {
  width: "100%",
  height: "100%",
};

const dashStroke = {
  hasStroke: true,
  strokeDasharray: "3 3",
};

export const IssuingAtmTransactionVolumeCount = ({data}:IssuingFaileSuccessProps) => {
  const lines = [
    {
      type: "monotone",
      dataKey: "successValue",
      stroke: "#C6F8DF",
      activeDot: { r: 8 },
    },
    {
      type: "monotone",
      dataKey: "failedValue",
      stroke: "#DC4437",
      activeDot: { r: 8 },
    },
  ];
  return (
    <>
      <IssuingLineChart
        lines={lines}
        width={500}
        height={300}
        dataKey="duration"
        data={data}
        container={containerType}
        stroke={dashStroke}
      />
    </>
  );
};

export const TransactionTypeBarChart = () => {
    const { userDetail } = useContext(AuthContext);
    const { transactionPeriod, selectedTenantCode } = useContext(StatsContext);

    const isSuperAdmin = userDetail?.role.name === appRoles.superAdmin

    let transactionCountVolumeUrl = getUrlForSuperadminORBankAdmin(apiUrlsv1.issuingVolumeStatus, selectedTenantCode )

    transactionCountVolumeUrl = `${transactionCountVolumeUrl}/volume/channel`
    transactionCountVolumeUrl = isSuperAdmin && (selectedTenantCode == '0'|| typeof selectedTenantCode === 'undefined')? `${transactionCountVolumeUrl}?dateRange=${transactionPeriod.toUpperCase()}&page=${0}&size=${20}` :`${transactionCountVolumeUrl}?tenantCode=${selectedTenantCode}&dateRange=${transactionPeriod.toUpperCase()}`
    const { isValidating, mutate, data, error } = useSWR(
        userDetail && userDetail?.role.name ? transactionCountVolumeUrl : ''
      );

  return (
    <>
      <IssuingBarChart
        labelY="Volume of each transaction types issuing"
        labelX={`${transactionPeriod} total ATM transaction metrics`}
        data={data?.response[0]}
      />
    </>
  );
};

export const BarChartHorizontal = () => {
    const { userDetail } = useContext(AuthContext);
    const { transactionPeriod, selectedTenantCode } = useContext(StatsContext)
    const tenantTopFailuredResons = apiUrlsv1.issuingFailedReasonsTenant
    let transactionCountVolumeUrl = getUrlForSuperadminORBankAdmin(apiUrlsv1.issuingVolumeStatus, selectedTenantCode )
    const isSuperAdmin = userDetail?.role.name === appRoles.superAdmin

    transactionCountVolumeUrl = `${transactionCountVolumeUrl}/failed-transaction`
    transactionCountVolumeUrl = isSuperAdmin && (selectedTenantCode == '0'|| typeof selectedTenantCode === 'undefined')? `${transactionCountVolumeUrl}?dateRange=${transactionPeriod.toUpperCase()}&page=${0}&size=${20}`:`${tenantTopFailuredResons}?tenantCode=${selectedTenantCode}&dateRange=${transactionPeriod.toUpperCase()}`
    const { isValidating, mutate, data, error } = useSWR(
        userDetail && userDetail?.role.name ? transactionCountVolumeUrl : ''
      );
      
  return (
    <Flex width="50%" sx={chartContainerSx}>
      <Text variant="chart-header">{transactionPeriod} transaction failure reasons</Text>
      <IssuingBarChartHorizontal
        labelX="Transaction volume"
        labelY="Top 15 transaction response codes"
        data={data?.response}
      />
    </Flex>
  );
};

export const IssuingTranValueChart = () => {
    const { userDetail } = useContext(AuthContext);
    const { transactionPeriod, selectedTenantCode } = useContext(StatsContext)
    let transactionCountVolumeUrl = getUrlForSuperadminORBankAdmin(apiUrlsv1.issuingVolumeStatus, selectedTenantCode )

    const isSuperAdmin = userDetail?.role.name === appRoles.superAdmin

    transactionCountVolumeUrl = `${transactionCountVolumeUrl}/volume`
    transactionCountVolumeUrl = isSuperAdmin && (selectedTenantCode == '0'|| typeof selectedTenantCode === 'undefined')? `${transactionCountVolumeUrl}?dateRange=${transactionPeriod.toUpperCase()}&page=${0}&size=${20}`:`${transactionCountVolumeUrl}?tenantCode=${selectedTenantCode}&dateRange=${transactionPeriod.toUpperCase()}`

    // transactionCountVolumeUrl = `${transactionCountVolumeUrl}/volume?dateRange=${transactionPeriod.toUpperCase()}&page=${0}&size=${20}`
    const { isValidating, mutate, data, error } = useSWR(
        userDetail && userDetail?.role.name ? transactionCountVolumeUrl : ''
      );

  return (
    <Flex height="100%" width="50%" sx={chartContainerSx}>
      <Text variant="chart-header">
          {transactionPeriod} breakdown of issuing transaction value
        </Text>

      <IssuingBarLineChart
        barSize={20}
        data={data?.response?.transactionDetails}
      />
    </Flex>
  );
};

export const IssuingTranVolumeChart = () => {
    const { userDetail } = useContext(AuthContext);
    const { transactionPeriod, selectedTenantCode } = useContext(StatsContext)
    let transactionCountVolumeUrl = getUrlForSuperadminORBankAdmin(apiUrlsv1.issuingVolumeStatus, selectedTenantCode );

    const isSuperAdmin = userDetail?.role.name === appRoles.superAdmin

    transactionCountVolumeUrl = `${transactionCountVolumeUrl}/volume`
    transactionCountVolumeUrl = isSuperAdmin && (selectedTenantCode == '0'|| typeof selectedTenantCode === 'undefined')? `${transactionCountVolumeUrl}?dateRange=${transactionPeriod.toUpperCase()}&page=${0}&size=${20}`:`${transactionCountVolumeUrl}?tenantCode=${selectedTenantCode}&dateRange=${transactionPeriod.toUpperCase()}`

    // transactionCountVolumeUrl = `${transactionCountVolumeUrl}/volume?dateRange=${transactionPeriod.toUpperCase()}&page=${0}&size=${20}`
    const { isValidating, mutate, data, error } = useSWR(
        userDetail && userDetail?.role.name ? transactionCountVolumeUrl : ''
      );

      const totalFailedTransaction = data && data?.response?.transactionDetails.length > 0 ? sumBy(data?.response?.transactionDetails,(count:any) => count?.count):0

  return (
    <Flex width="50%" sx={chartContainerSx}>
      <HStack justifyContent="space-between">
      <Text variant="chart-header">{transactionPeriod} count of Issued cards</Text>
        <Tag>
          Total: <strong>{totalFailedTransaction}</strong>
        </Tag>
      </HStack>
      <IssuingLineChartSingle
        data={data?.response?.transactionDetails}
        tickCount={6}
        type="number"
        interval={0}
      />
    </Flex>
  );
};
