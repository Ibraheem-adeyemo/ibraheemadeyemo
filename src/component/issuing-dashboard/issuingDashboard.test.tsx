import React from 'react';
import { IssuingTranValueChart, IssuingTranVolumeChart } from './IssuingAtmTransactionVolumeCount'
import { FailedAndSuccessfulChart } from './'
import { render, screen } from '@testing-library/react';
import { shallow, configure, mount } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom'

configure({adapter: new Adapter()})

const { ResizeObserver } = window;

global.ResizeObserver = jest.fn().mockImplementation(() => ({
        observe: jest.fn(),
        unobserve: jest.fn(),
        disconnect: jest.fn(),
    }))

describe('IssuingTransactionValueChart', ()=>{
    const wrapper = shallow(<IssuingTranValueChart />)
    it('should match the snapshot', ()=>{
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('IssuingTranValueChart component rendered properly', ()=>{
        render(<IssuingTranValueChart />)
        const welcomStatement = screen.getByText(/breakdown of issuing/, {exact: false})
        expect(welcomStatement).toBeInTheDocument()
    })    

    test('IssuingTranVolumeChart component rendered properly', ()=> {
        render(<IssuingTranVolumeChart />)
        const activeCardChart = screen.getByText(/count of active cards/, {exact: false})
        expect(activeCardChart).toBeInTheDocument()
    });
})

describe('IssuingTranVolumeChart', ()=>{
    it('IssuingTranVolumeChart component rendered properly', ()=>{
        const wrapper = shallow(<IssuingTranVolumeChart />)
        // const wrapperContext = wrapper.context();
        const text = wrapper.find('.text').text()
        expect(text).toEqual(' count of active cards')
    })
})

describe('FailedAndSuccessfulChart', ()=>{

    
    beforeEach(() =>{
        //delete window?.ResizeObserver;
        // const wrapper = mount(<FailedAndSuccessfulChart />)
    })


    it('should match the snapshot', ()=>{
        const tree = renderer.create(<FailedAndSuccessfulChart />).toJSON()
        expect(tree).toMatchSnapshot()
    })
    it('Displays Total transaction', ()=>{
        const wrapper = shallow(<FailedAndSuccessfulChart />).dive();
        const originalContext = React.useContext;
        const mockReactUseContext = jest.fn().mockReturnValue('daily')
        React.useContext = mockReactUseContext
        const text = wrapper.find('.totalTrans').text()
        expect(text).toEqual('Total  transaction')
    })

    it('display failed transaction', ()=>{
        const wrapper = shallow(<FailedAndSuccessfulChart />)
        const text = wrapper.find('.failedTransaction').text()
        expect(text).toEqual('Failed transaction')
    })

    it('display success transaction', ()=>{
        const wrapper = shallow(<FailedAndSuccessfulChart />)
        const text = wrapper.find('.successTransaction').text()
        expect(text).toEqual('Successful transaction')
    })
})