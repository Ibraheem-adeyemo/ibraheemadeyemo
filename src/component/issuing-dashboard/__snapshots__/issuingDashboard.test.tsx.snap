// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuingTransactionValueChart should match the snapshot 1`] = `"<div class="css-1vgdsvy"><div class="chakra-skeleton css-1verwtw"></div></div><div class="css-1vgdsvy"><div class="chakra-skeleton css-1verwtw"></div></div><div class="css-1vgdsvy"><div class="chakra-skeleton css-1verwtw"></div></div><div class="css-1vgdsvy"><div class="chakra-skeleton css-1verwtw"></div></div><div class="css-1vgdsvy"><div class="chakra-skeleton css-1verwtw"></div></div><div class="css-1vgdsvy"><div class="chakra-skeleton css-1verwtw"></div></div>"`;
